t/socket-tcp-getoption.t

```diff
@@ -0,0 +1,288 @@
+            if not ok then
+                ngx.say("failed to connect: ", err)
+                return
+            end
+
+            local val, err = sock:getoption("reuseaddr")
+            if not val then
+                ngx.say("getoption reuseaddr failed: ", err)
+                return
+            end
+
+            ngx.say("reuseaddr: ", val)
+
+            sock:close()
+        }
+    }
+--- request
+GET /t
+--- response_body_like eval
+qr/reuseaddr: \d+/
+--- no_error_log
+[error]
+
+
+
+=== TEST 6: get tcp-nodelay.
+--- config
+    set $port $TEST_NGINX_SERVER_PORT;
+
+    location /t {
+        content_by_lua_block {
+            require "resty.core.socket"
+
+            local port = ngx.var.port
+            local sock = ngx.socket.tcp()
+            local ok, err = sock:connect("127.0.0.1", port)
+            if not ok then
+                ngx.say("failed to connect: ", err)
+                return
+            end
+
+            local val, err = sock:getoption("tcp-nodelay")
+            if not val then
+                ngx.say("getoption tcp-nodelay failed: ", err)
+                return
+            end
+
+            ngx.say("tcp-nodelay: ", val)
+
+            sock:close()
+        }
+    }
+--- request
+GET /t
+--- response_body_like eval
+qr/tcp-nodelay: \d+/
+--- no_error_log
+[error]
+
+
+
+=== TEST 7: get sndbuf.
+--- config
+    set $port $TEST_NGINX_SERVER_PORT;
+
+    location /t {
+        content_by_lua_block {
+            require "resty.core.socket"
+
+            local port = ngx.var.port
+            local sock = ngx.socket.tcp()
+            local ok, err = sock:connect("127.0.0.1", port)
+            if not ok then
+                ngx.say("failed to connect: ", err)
+                return
+            end
+
+            local val, err = sock:getoption("sndbuf")
+            if not val then
+                ngx.say("getoption sndbuf failed: ", err)
+                return
+            end
+
+            ngx.say("sndbuf: ", val)
+
+            sock:close()
+        }
+    }
+--- request
+GET /t
+--- response_body_like eval
+qr/sndbuf: \d+/
+--- no_error_log
+[error]
+
+
+
+=== TEST 8: get rcvbuf.
+--- config
+    set $port $TEST_NGINX_SERVER_PORT;
+
+    location /t {
+        content_by_lua_block {
+            require "resty.core.socket"
+
+            local port = ngx.var.port
+            local sock = ngx.socket.tcp()
+            local ok, err = sock:connect("127.0.0.1", port)
+            if not ok then
+                ngx.say("failed to connect: ", err)
+                return
+            end
+
+            local val, err = sock:getoption("rcvbuf")
+            if not val then
+                ngx.say("getoption rcvbuf failed: ", err)
+                return
+            end
+
+            ngx.say("rcvbuf: ", val)
+
+            sock:close()
+        }
+    }
+--- request
+GET /t
+--- response_body_like eval
+qr/rcvbuf: \d+/
+--- no_error_log
+[error]
```

Comment:
Needs test cases for the cases that the C function setsockopt() returns errors and we can get the system error messages for the failures.

Reply:
it hard to create a test case for the C function for parameters check in lua and no socket parameters can be got in lua, so i test it manual by code the parameters to wrong value in C function.

Reply:
One way to test this is to introduce a special API to allow overriding the underlying fd to an invalid number (but not -1), then setsockopt will return an error.